R/grid.R

```diff
@@ -0,0 +1,295 @@
+    )
+  )
+}
+
+#' Define a template of a CSS grid
+#'
+#' @param default
+#' Template for desktop:
+#' list(areas = [data.frame of strings],
+#'      rows_height = [vector of strings],
+#'      cols_width = [vector of strings])
+#' @param mobile
+#' Template for mobile:
+#' list(areas = [data.frame of strings],
+#'      rows_height = [vector of strings],
+#'      cols_width = [vector of strings])
+#'
+#' @return
+#' list(template = [string], area_names = [vector of strings])
+#'
+#' template - contains template that can be formatted with glue::glue() function
+#'
+#' area_names - contain all unique area names used in grid definition
+#'
+#' @details
+#' Example usage:
+#' \preformatted{
+#'   myGrid <- grid_template(default = list(
+#'     areas = rbind(
+#'       c("header", "header", "header"),
+#'       c("menu",   "main",   "right1"),
+#'       c("menu",   "main",   "right2")
+#'     ),
+#'     rows_height = c("50px", "auto", "100px"),
+#'     cols_width = c("100px", "2fr", "1fr")
+#'   ))
+#'
+#'   display_grid(myGrid)
+#' }
+#' Second example:
+#' \preformatted{
+#'   subGrid <- grid_template(default = list(
+#'     areas = rbind(
+#'       c("top_left", "top_right"),
+#'       c("bottom_left", "bottom_right")
+#'     ),
+#'     rows_height = c("50\%", "50\%"),
+#'     cols_width = c("50\%", "50\%")
+#'   ))
+#'
+#'   display_grid(subGrid)
+#' }
+#'
+#' @export
+#'
+grid_template <- function(
+  default = list(areas = "", rows_height = c(), cols_width = c()),
+  mobile = list(areas = "", rows_height = c(), cols_width = c())) {
+  # TODO: Support for mobile grid version. Only `default` argument is used right now.
+
+  area_names <- default$areas %>% as.vector %>% unique
+  css_grid_template_areas <- data_frame_to_css_grid_template_areas(default$areas)
+
+  grid_template <- shiny::tags$div(
+    style = grid_container_css(css_grid_template_areas, default$rows_height, default$cols_width),
+    shiny::tagList(list_of_area_tags(area_names))
+  ) %>% htmltools::renderTags(.)
+
+  return(list(template = grid_template$html, area_names = area_names))
+}
+
+#' Format string template (that represents HTML template) with custom CSS styles.
+#'
+#' @param html_template string
+#' @param area_names vector of strings
+#' @param container_style string
+#' @param area_styles list of strings
+#'
+#' @return string
+#'
+#' @details This is a helper function used in grid()
+#'
+apply_custom_styles_to_html_template <- function(html_template = "",
+                                                 area_names = c(),
+                                                 container_style = "",
+                                                 area_styles = list()) {
+  custom_styles <- list(custom_style_grid_container = container_style)
+  for (area in area_names) {
+    custom_styles[[glue::glue("custom_style_grid_area_{area}")]] <- ifelse(
+      area %in% names(area_styles), area_styles[[area]], "")
```

Comment:
I'm surprised that the checks have passed. I thought it would complain about lack of import magrittr here (`%in%`), but anyway if it works with `--as-cran` checks we may leave it as is

Reply:
`%in%` is base R operator for checking if vector contains an element, it's not imported from magrittr :)